### PR TITLE
Add fromXContent to InternalNestedIdentity

### PR DESCRIPTION
**Pull request by cbuescher**
This adds a fromXContent method and unit test to InternalNestedIdentity so we can parse it as part of a search response. This is part of the preparation for parsing search responses on the client side.

### CODE REVIEW

**Review by martijnvg — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java [hunk 1]
```diff
@@ -795,12 +816,42 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             return builder;
         }
 
-        public static class Fields {
+        private static final ConstructingObjectParser<InternalNestedIdentity, ParseFieldMatcherSupplier> PARSER
+                = new ConstructingObjectParser<>("nested_identity",
+                        a -> new InternalNestedIdentity((String) a[0], (int) a[1], (InternalNestedIdentity) a[2]));
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: can we call the variable something else than `a`?

**Review by javanna — APPROVED** [review]
LGTM, left a couple of minors

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java [hunk 2]
```diff
@@ -557,7 +564,9 @@ public void readFrom(StreamInput in) throws IOException {
         score = in.readFloat();
         id = in.readOptionalText();
         type = in.readOptionalText();
-        nestedIdentity = in.readOptionalStreamable(InternalNestedIdentity::new);
+        if (in.readBoolean()) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I want to introduce a ctor that takes a StreamInput, but to do that I need to read the bool that signals if the object was written outside. I changed this a bit furhter now so InternalNestedIdentity implements `Writeable`. I agree agree we generally don't want big changes here, but I think in terms of wire protocol this is equivalent and to me makes the code easier to read.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Changed to `ctorArgs`.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java [hunk 3]
```diff
@@ -557,7 +564,9 @@ public void readFrom(StreamInput in) throws IOException {
...
+            nestedIdentity = in.readOptionalWriteable(InternalNestedIdentity::new);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I would not expect the if here, right?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java [hunk 4]
```diff
@@ -747,7 +756,12 @@ public InternalNestedIdentity(String field, int offset, InternalNestedIdentity c
             this.child = child;
         }
 
-        InternalNestedIdentity() {
+        InternalNestedIdentity(StreamInput in) throws IOException {
+            field = in.readOptionalText();
+            offset = in.readInt();
+            if (in.readBoolean()) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: readOptionalWriteable?

**Review by javanna — COMMENTED** [review]
left comments on the serialization changes. I would not to do the same with other classes in the future as part of this fromXContent effort.

**Comment by cbuescher** [comment]
On 5.x with ae9088ccfaf3f02c2327867071473ce08aff419d

**Pull request merged**